teps/0080-taskrun-prepost-steps.md

```diff
@@ -0,0 +1,163 @@
+---
+status: proposed
+title: 'TaskRun: Pre/Post Steps'
+creation-date: '2021-08-18'
+last-updated: '2021-08-18'
+authors:
+- '@mattmoor'
+---
+
+# TEP-0080: TaskRun: Pre/Post Steps
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+This TEP allows users to partition their list of steps into three distinct phases
+(naming is "to-be-bikeshed"):
+ - `pre-steps`: a section of the TaskRun execution prior to the "real" work.  This
+ would generally consist of "setup" and "download" related work.
+ - `steps`: what we have today.
+ - `post-steps`: a section of the TaskRun execution posterior to the "real" work.
+ This would generally consist of "tear-down" and "upload" related work.
```

Comment:
This point is handled by [TEP-0040](https://github.com/tektoncd/community/blob/main/teps/0040-ignore-step-errors.md), aka be able to run steps no matter what was the result before, doesn't it ?

Reply:
I think you are assuming (because I haven't stated as much!) that this runs like a `finally` block (e.g. regardless of success/failure).  I could go either way on this, but I see (at least) three possibilities:

1. `post-steps` have `finally` semantics, which would enable TEP-0040 like cleanup logic.
2. `post-steps` have normal semantics, which would complement TEP-0040 in that folks could use `post-steps` like a `finally` block, if they so choose.
3. `post-steps` augments TEP-0040, so that folks can write `onError: post-steps` (or something) to indicate that when a step fails, that execution should flow to the post-steps (perhaps without executing the intervening steps).

IMO `2.` and `3.` leave this as non-conflicting/complementary to TEP-0040, where `1.` this might subsume/replace TEP-0040.

Reply:
Honestly, when I first wrote this `2.` is what I had in mind, but I think you read it as `1.`, so it would be good to flesh this out and hopefully that shows why I wasn't thinking of this as conflicting with TEP-0040 😅